# Project Speed Dating

## Project Target

Tinder have experienced a decrease in the number of matches and are trying to understand __what makes people interested into each other__. 
They have provided us a dataset to understand the drivers for a second date

## Dataset

Let's open and read the dataset

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
import statistics

df= pd.read_csv('Speed+Dating+Data.csv',encoding='iso-8859-1')
df.shape

(8378, 195)

There are 8,378 lines representing one subject participating to a date and rating it. Let's keep in mind that each date includes 2 persons and is represented by 2 lines in the dataset. Let's check the column of the dataset

In [2]:
for column in df.columns:
    print(column)

iid
id
gender
idg
condtn
wave
round
position
positin1
order
partner
pid
match
int_corr
samerace
age_o
race_o
pf_o_att
pf_o_sin
pf_o_int
pf_o_fun
pf_o_amb
pf_o_sha
dec_o
attr_o
sinc_o
intel_o
fun_o
amb_o
shar_o
like_o
prob_o
met_o
age
field
field_cd
undergra
mn_sat
tuition
race
imprace
imprelig
from
zipcode
income
goal
date
go_out
career
career_c
sports
tvsports
exercise
dining
museums
art
hiking
gaming
clubbing
reading
tv
theater
movies
concerts
music
shopping
yoga
exphappy
expnum
attr1_1
sinc1_1
intel1_1
fun1_1
amb1_1
shar1_1
attr4_1
sinc4_1
intel4_1
fun4_1
amb4_1
shar4_1
attr2_1
sinc2_1
intel2_1
fun2_1
amb2_1
shar2_1
attr3_1
sinc3_1
fun3_1
intel3_1
amb3_1
attr5_1
sinc5_1
intel5_1
fun5_1
amb5_1
dec
attr
sinc
intel
fun
amb
shar
like
prob
met
match_es
attr1_s
sinc1_s
intel1_s
fun1_s
amb1_s
shar1_s
attr3_s
sinc3_s
intel3_s
fun3_s
amb3_s
satis_2
length
numdat_2
attr7_2
sinc7_2
intel7_2
fun7_2
amb7_2
shar7_2
attr1_2
sinc1_2
intel1_2
fun1_2
amb1_2
shar1_2
attr4_2
sinc4_2
intel4_2
fun4_2
amb

There seems to have a large number of columns. By using the dataset description provided by Tinder, here are the different category
- __Subject__ : id, gender
- __Partner__  : id, age, race
- __Wave__ : Wave number, position of the date in the wave
- __Date__ : decision of partner and subject, match, common interest, final date
- __Scorecards__
    - Rating done by the partner and the subject on the partner of the opposite sex
    - Category : Attractive, Sincere, Intelligent, Fun, Ambitious, Shared Hobbies, 
- __Survey__ : Different survey has been performed at different time and with different repartition by wave
    - Ranking of the previous criteria on different question
    - Different question : what you are looking for? What do you think other from the same sex are looking for? What the opposite sex is looking for? How do you rate yourself? 
    - Different timing : surveys done during the date, few days after and few weeks after
    - Additional survey on studen done on only one date

## Scorecard Analysis

As a first approach, let's check the average score of the partner based on the final decision

In [3]:
attributes = ['attr','sinc','intel','fun','amb','shar']
df_positive = df[df['dec']==1]
df_positive_mean = df_positive.groupby(['gender'])[attributes].mean().reset_index()
df_positive_mean

gender      attr      sinc     intel       fun       amb      shar
0       0  7.072787  7.632566  7.932164  7.381710  7.374647  6.583885
1       1  7.448739  7.568112  7.621498  7.296134  6.995148  6.371681

In [4]:

y_female = df_positive_mean.iloc[0,1:]
y_male = df_positive_mean.iloc[1,1:]
column = df_positive_mean.columns[1:]

fig = go.Figure()

fig.add_trace(
    go.Bar(
        x=column,
        y = y_female,
        name='Female',
        marker_color='indianred',
        text=round(y_female,2),
        textposition='auto'
    )
)

fig.add_trace(
    go.Bar(
        x=column,
        y = y_male,
        name='Male',
        marker_color='lightsalmon',
        text=round(y_male,2),
        textposition='auto'
    )
)



fig.update_layout(
    title = go.layout.Title(text='Average rating of each category depending in the gender when the decision if positive', x=0.5),
    width=800
)
fig.show()



With this quick graph, we can see that
- All criteria have relatvely high average, as there all except shared interest above 7
- Intelligence has the highest average for both male and female, so it seems to be an important criteria 
- Average attractivness has an higher average for male than for female. It seems that attractivness has an higher role for male
- On the other hand, Ambitious seems to have an higher importance for female


However, let's compare the average rate when the decision is negative 

In [15]:
attributes = ['attr','sinc','intel','fun','amb','shar','like','prob','met']
df_attributes_mean = df.groupby(['gender','dec'])[attributes].mean().reset_index()
df_attributes_mean

gender  dec      attr      sinc     intel       fun       amb      shar  \
0       0    0  5.231626  6.780189  7.157363  5.613764  6.704489  4.695865   
1       0    1  7.072787  7.632566  7.932164  7.381710  7.374647  6.583885   
2       1    0  5.531739  6.952563  6.977426  5.798275  6.237808  4.747230   
3       1    1  7.448739  7.568112  7.621498  7.296134  6.995148  6.371681   

       like      prob       met  
0  5.286137  4.742840  0.965130  
1  7.169743  6.004980  0.929521  
2  5.344513  4.497596  0.898551  
3  7.262097  5.940833  0.996381

Let's put this value in some graph to have a better understanding of the data

In [18]:

for attribute in attributes:
    fig = go.Figure()
    gender = df_attributes_mean['gender'].unique()

    fig.add_trace(
        go.Bar(
            x=gender,
            y = df_attributes_mean[df_attributes_mean['dec']==1][attribute],
            name='Positive Decision',
            marker_color='indianred',
            text=round(df_attributes_mean[df_attributes_mean['dec']==1][attribute],2),
            textposition='auto'
        )
    )

    fig.add_trace(
        go.Bar(
            x=gender,
            y = df_attributes_mean[df_attributes_mean['dec']==0][attribute],
            name='Negative Decision',
            marker_color='lightsalmon',
            text=round(df_attributes_mean[df_attributes_mean['dec']==0][attribute],2),
            textposition='auto'
        )
    )



    fig.update_layout(
        title = go.layout.Title(text='Average rating of {} depending on the gender and on the final decision'.format(attribute), x=0.5),
        width=800
    )
    fig.show()


What we can deduct from this graph
- All the categories are important for the choice as the average is always higher when the decision if positive
- However, there are discrepencies in the difference of average between a positive and negative decision
    - The Intelligence rating of the negative decision is almost at 7, which shows that a high number of partner where not choosen despite being rated high on Intelligence
    - On the other hand, there is a substantial gap in average between a positive and negative decision for Attractivness, Fun and Shared Interest. It means that if the score is low, there are high change that the partner will be rejectet


Overall, it seems that these 3 factors are the main factors of decision: Attractivness, Fun, Intelligence and Shared Interest

Let's see the distribution of the rating  confirm it

In [17]:
attributes = ['attr','sinc','intel','fun','amb','shar']

for attribute in attributes:
    for i in [0,1]:
        fig = px.histogram(
            df[(df['dec']==1)&(df['gender']==i)],
            x=attribute,
            color_discrete_sequence=['lightgreen'],
            nbins=10,
            title = 'Distribution of {} ranking when the decision is positive - {}'.format(attribute, 'male' if i==1 else 'female'),
        )
        fig.show()

        fig = px.histogram(
            df[(df['dec']==0)&(df['gender']==i)],
            x=attribute,
            color_discrete_sequence=['lightsalmon'],
            nbins=10,
            title = 'Distribution of {} ranking when the decision is negative - {}'.format(attribute, 'male' if i==1 else 'female')
        )
        fig.show()

Previous histograms are confirming the previous analyssis
- Intelligence, Sincerity and Ambition have a similar distribution for positive and negative decision, therefore are not the main drive
- On the other hand, Attractivness, Fun and Shared Activies are clearly showing different distribution, which shows that these criteria are driving the decision

Let's see if the is in line with the answer of the survey

## Survey Analysis

Let's review if the survey can be used. let's check if there are a lot of NaN in the dataframe

In [86]:
survey = []
for column in df.filter(regex='intel').columns:
    survey.append(column[5:])
survey.sort()
survey = survey[1:len(survey)-1]

columns_survey=[]
for columns in df.columns:
    for test in survey:
        if test in columns:
            columns_survey.append(columns)
            break

nan_counts = (df[columns_survey].isna().mean()*100).sort_values().reset_index()
nan_counts.columns = ['Column', 'NaN_Count']
nan_counts


fig = px.bar(
    nan_counts,
    x='Column',
    y='NaN_Count'   
)
fig.show()
print(survey)
print(columns_survey)

['1_1', '1_2', '1_3', '1_s', '2_1', '2_2', '2_3', '3_1', '3_2', '3_3', '3_s', '4_1', '4_2', '4_3', '5_1', '5_2', '5_3', '7_2', '7_3']
['attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1', 'attr4_1', 'sinc4_1', 'intel4_1', 'fun4_1', 'amb4_1', 'shar4_1', 'attr2_1', 'sinc2_1', 'intel2_1', 'fun2_1', 'amb2_1', 'shar2_1', 'attr3_1', 'sinc3_1', 'fun3_1', 'intel3_1', 'amb3_1', 'attr5_1', 'sinc5_1', 'intel5_1', 'fun5_1', 'amb5_1', 'attr1_s', 'sinc1_s', 'intel1_s', 'fun1_s', 'amb1_s', 'shar1_s', 'attr3_s', 'sinc3_s', 'intel3_s', 'fun3_s', 'amb3_s', 'attr7_2', 'sinc7_2', 'intel7_2', 'fun7_2', 'amb7_2', 'shar7_2', 'attr1_2', 'sinc1_2', 'intel1_2', 'fun1_2', 'amb1_2', 'shar1_2', 'attr4_2', 'sinc4_2', 'intel4_2', 'fun4_2', 'amb4_2', 'shar4_2', 'attr2_2', 'sinc2_2', 'intel2_2', 'fun2_2', 'amb2_2', 'shar2_2', 'attr3_2', 'sinc3_2', 'intel3_2', 'fun3_2', 'amb3_2', 'attr5_2', 'sinc5_2', 'intel5_2', 'fun5_2', 'amb5_2', 'attr1_3', 'sinc1_3', 'intel1_3', 'fun1_3', 'amb1_3', 'shar1_3', 'attr7_3',

In [97]:
surveys = []
for column in df.filter(regex='intel').columns:
    surveys.append(column[5:])
surveys.sort()
surveys = survey[1:len(surveys)-1]


columns_survey=[]
for columns in df.columns:
    for test in surveys:
        if test in columns:
            columns_survey.append(columns)
            break

max_nan = {i: 0 for i in surveys}
print(max_nan)

for column in df[columns_survey].columns:
    for item in survey:
        if item in column:
            if max_nan[item]<df[column].isna().mean()*100:
                max_nan[item]=df[column].isna().mean()*100


max_nan.pop('2_1')  #2_1 is What do you think the opposite sex looks for in a date?. Cannot be used
max_nan.pop('3_1')  #3_1 is How do you think you measure up?. Cannot be used
max_nan.pop('3_2')  #3_1 is How do you think you measure up?. Cannot be used

print(min(max_nan,key=max_nan.get),max_nan[min(max_nan,key=max_nan.get)])

{'1_2': 0, '1_3': 0, '1_s': 0, '2_1': 0, '2_2': 0, '2_3': 0, '3_1': 0, '3_2': 0, '3_3': 0, '3_s': 0, '4_1': 0, '4_2': 0, '4_3': 0, '5_1': 0, '5_2': 0, '5_3': 0, '7_2': 0, '7_3': 0}
1_2 11.136309381714012


The survey to be used it the 1_2 has there are almost 90% answer and the question is in line with our research : We want to know what you look for in the opposite sex. 

However, there are 2 different ratings that cannot be compared. Let's find which are the wave with the most partipant

In [108]:
df[df['wave'].isin([i for i in range(1,6)]+[i for i in range(10,22)])].shape

(6816, 196)

In [126]:
df_survey = df[df['wave'].isin([i for i in range(1,6)]+[i for i in range(10,22)])]
column_survey = [item for item in df_survey.columns if '1_1' in item]
for column in column_survey:
    df_survey = df_survey[df_survey[column].notna()]

df_survey_attributes = df_survey.groupby(['gender'])[column_survey].mean()
df_survey_attributes

attr1_1    sinc1_1   intel1_1     fun1_1     amb1_1    shar1_1
gender                                                                  
0       18.648623  18.432483  21.512705  17.090132  12.018897  12.264193
1       28.925728  16.240439  19.591681  17.650809   7.498773  10.258607

In [129]:
y_female = df_survey_attributes.iloc[0,0:]
y_male = df_survey_attributes.iloc[1,0:]
column = column_survey

fig = go.Figure()

fig.add_trace(
    go.Bar(
        x=column,
        y = y_female,
        name='Female',
        marker_color='indianred',
        text=round(y_female,2),
        textposition='auto'
    )
)

fig.add_trace(
    go.Bar(
        x=column,
        y = y_male,
        name='Male',
        marker_color='lightsalmon',
        text=round(y_male,2),
        textposition='auto'
    )
)



fig.update_layout(
    title = go.layout.Title(text='Average rating of each category depending in the gender', x=0.5),
    width=800
)
fig.show()

## Deep-Dive Analysis


Based on the exiting dataset, we can deep-dive and check the influence of the followign factors
- Age
- Race
- Position inside the wave

### Position inside the wave

In [9]:


df_by_order = df[df['gender']==0].groupby(['order'])['dec'].agg(['sum','count']).reset_index()
print('for female, the average positive answer is {}'.format(round(df_by_order['sum'].sum()/df_by_order['count'].sum(),2)))
df_by_order['%_Positive'] = 0
df_by_order['%_Positive'] = df_by_order['sum']/df_by_order['count']
fig = px.scatter(
    df_by_order,
    x='order',
    y='%_Positive',
    trendline='ols'
)
fig.show()



fig = px.bar(
    df_by_order,
    x='order',
    y='count',
)
fig.show()


df_by_order = df[df['gender']==1].groupby(['order'])['dec'].agg(['sum','count']).reset_index()
print('for male, the average positive answer is {}'.format(round(df_by_order['sum'].sum()/df_by_order['count'].sum(),2)))

df_by_order['%_Positive'] = 0
df_by_order['%_Positive'] = df_by_order['sum']/df_by_order['count']
fig = px.scatter(
    df_by_order,
    x='order',
    y='%_Positive',
    trendline='ols'
)



fig.show()

fig = px.bar(
    df_by_order,
    x='order',
    y='count',
)
fig.show()

print(df_by_order['sum'].sum()/df_by_order['count'].sum())


for female, the average positive answer is 0.37


for male, the average positive answer is 0.47


0.4742489270386266


With this analysis, we can see that the order has an important role : 
- The first order seems to have a higher positive answer both for male and female. Seems like the first round has a positive impact
- On the other, the second order seems to be one of the order with the lowest positive ratio. Their could be a phsychological bias that would decrease the chance of sucess when the first is positive
- The ration of positive answer slightly decrease during the wave

In [10]:
df_2nd_position = df[df['order'].isin([1,2])][['iid','order','dec']]
df_2nd_position['1r_dec']=0
order1_dec =df_2nd_position[df_2nd_position['order']==1][['iid','dec']].set_index('iid')
order1_dec = order1_dec.copy()

df_2nd_position.loc[df_2nd_position['order']==2,'1r_dec'] = df_2nd_position.loc[df_2nd_position['order'] == 2, 'iid'].map(order1_dec['dec'])
df_2nd_position = df_2nd_position.reset_index(drop=True)
df_2nd_position



df_2nd_position.groupby(['order','1r_dec'])['dec'].agg(['sum','count']).reset_index()


order  1r_dec  sum  count
0      1       0  275    551
1      2       0   79    276
2      2       1  138    275

### Age

In the dataset, there is for the majority of the subjects and partner their age. Let's check how the age_gap is impacting the final decision

In [11]:
df['age_gap'] = df['age_o'] - df['age']

print('male')
df_age_gap = df[df['gender']==1].groupby(['age_gap'])['dec'].agg(['sum','count']).reset_index()
df_age_gap['ratio'] = df_age_gap['sum'] / df_age_gap['count']
df_age_gap=df_age_gap[df_age_gap['count']>30]

fig = px.scatter(
    df_age_gap,
    x='age_gap',
    y='ratio'
)
fig.show()

fig = px.bar(
    df_age_gap,
    x='age_gap',
    y='count'
)
fig.show()

print('female')

df_age_gap = df[df['gender']==0].groupby(['age_gap'])['dec'].agg(['sum','count']).reset_index()
df_age_gap['ratio'] = df_age_gap['sum'] / df_age_gap['count']
df_age_gap=df_age_gap[df_age_gap['count']>30]

fig = px.scatter(
    df_age_gap,
    x='age_gap',
    y='ratio'
)
fig.show()

fig = px.bar(
    df_age_gap,
    x='age_gap',
    y='count'
)
fig.show()

male


female
